core/modules/main/controllers/Main.php

```diff
@@ -3810,92 +3853,174 @@ public function runGetBackdropPartial(framework\Request $request)
+                }
             }
-            catch (\Exception $e)
+            else
             {
-                $status = 400;
-                $message = framework\Context::getI18n()->__('Could not find this issue');
+                return $this->return404(framework\Context::getI18n()->__('Could not find this issue'));
             }
         }
-        elseif ($request->hasParameter('issue_id'))
+        else
         {
-            $status = 400;
-            $message = framework\Context::getI18n()->__('Please provide an issue number');
+            return $this->return404(framework\Context::getI18n()->__('This project does not exist'));
         }
 
-        $searchfor = $request['searchfor'];
+        // Prepare response and return it.
+        $this->getResponse()->setHttpStatus(framework\Response::HTTP_STATUS_OK);
+
+        $parameters = [
+            'selected_project' => $this->selected_project,
+            'issue' => $issue,
+            'grouped_issues' => $grouped_issues
+        ];
+
+        return $this->renderJSON(array('content' => $this->getComponentHTML('main/findrelatedissues', $parameters)));
+    }
+
+    /**
+     * Find issues that might be duplicates of selected issue.
+     *
+     * @param \thebuggenie\core\framework\Request $request
+     */
+    public function runFindDuplicatedIssue(framework\Request $request)
+    {
+        $issue_id = $request['issue_id'];
+        $searchfor = trim($request['searchfor']);
 
-        if (mb_strlen(trim($searchfor)) < 3 && !is_numeric($searchfor) && mb_substr($searchfor, 0, 1) != '#')
+        // Verify request parameters.
+        if ($issue_id === null)
         {
-//                $status = 400;
-//                $message = framework\Context::getI18n()->__('Please enter something to search for (3 characters or more) %searchfor', array('searchfor' => $searchfor));
-            $issues = array();
-            $count = 0;
+            return $this->return400(framework\Context::getI18n()->__('Please provide an issue number'));
         }
-        else
+        else if ($searchfor === null or $searchfor === "")
         {
-            $this->getResponse()->setHttpStatus($status);
-            if ($status == 400)
+            return $this->return400(framework\Context::getI18n()->__('Please provide search text'));
+        }
+        // Valid project has been referenced, user has permissions
+        // to access it (preExecute performs access check already).
+        else if ($this->selected_project instanceof entities\Project)
+        {
+            $issue = entities\Issue::getB2DBTable()->selectById($issue_id);
+
+            if ($issue instanceof entities\Issue && $issue->hasAccess($this->getUser()))
             {
-                return $this->renderJSON(array('error' => $message));
-            }
+                // Try to get both exact match and text-based search.
+                $exact_match = entities\Issue::getIssue($searchfor);
```

Comment:
Lines 3978-4002 seems identical with lines 3883-3992 - could it be refactored into a function?

Reply:
Also, what's the reasoning behind not doing the sorting and grouping when presenting duplicate issue search results?

Reply:
The important difference between the 3883-3992 and 3978-4002 is in lines 3885 and 3979:

```
3885: $found_issues = entities\Issue::findIssuesByText($searchfor, null);
3979: $matched_issues = entities\Issue::findIssuesByText($searchfor, $this->selected_project);
```

So, when finding issues for adding relation the search is done over all the projects, while in case of finding a duplicate, only the current project is looked into.

I wasn't sure it would help much with deduplication, and it felt like it could reduce readability of code (kind of lots of wrapping around `Issue`'s method for finding by text). Also kind of going for a "three-strike" rule for deduplicating. Would you like me to create another method instead for this piece?

In case of duplicate issues not being sorted, the user is presented with a simple drop-down menu, plus all the issues are coming from the same project, so I figured there's not much point to grouping the issues beyond making sure to fit the exact match at top. Although, I can see some value in sorting the issues below it by issue ID. I guess the sorting could be done for those as well, and then the template would cycle through them in two loops.

That being said, if both of those changes are made, maybe then it makes sense to remerge the two actions again back into a single one. What do you think?

Reply:
Well, in that case I agree with you. I think your solution with splitting them up is better, and some duplication doesn't always hurt.

Reply:
Should I perhaps do a sort on issues by ID for duplicate issues?